specifications/environment-management.org

```diff
@@ -0,0 +1,121 @@
+* Environment Management: Scope
+  :PROPERTIES:
+  :CUSTOM_ID: environment-management-scope
+  :END:
+
+Specification of how snapcraft manages local and remote build
+environments and their lifecycle.
+
+This includes locally and remotely provisioned containers, VMs and build services such as Launchpad.
+
+* Environment
+  :PROPERTIES:
+  :CUSTOM_ID: environment
+  :END:
+
+** Identification
+   :PROPERTIES:
+   :CUSTOM_ID: identification
+   :END:
+
+*** Local
+    :PROPERTIES:
+    :CUSTOM_ID: local
+    :END:
+
+Unique identifier to lookup build instance from a local database.
+
+Format: =<project-name>-<provider>-<project-directory-hex-encoded>=
+
+Where:
+
+- =<project-directory-hex-encoded>= is an absolute path represented in hex, e.g. =str(Path(<project-path>).absolute()).encode().hex()=.
+
+- =<provider>= is the provider type (launchpad, multipass, lxd, host, etc.).
+
+** Build Instance Naming
+   :PROPERTIES:
+   :CUSTOM_ID: build-instance-naming
+   :END:
+
+Build instance will be named =snapcraft-<project_name>=. If name is unavailable, attempt to reuse it. Otherwise use =snapcraft-<project_id>=.
+
+** Build Instance Lifecycle
+   :PROPERTIES:
+   :CUSTOM_ID: build-instance-lifecycle
+   :END:
+
+*** Creation
+    :PROPERTIES:
+    :CUSTOM_ID: creation
+    :END:
+
+Creating an environment when running snapcraft inside a project:
+
+#+BEGIN_EXAMPLE
+    def is_stale(env):
+      elapsed = time.now() - env.last_used
+      return elapsed.days > 7
+
+    def is_environment_reusable(env):
+      return not env.project_path.exists() or env.is_stale():
+
+    def get_environment(project_name, project_id):
+      name = f"snapcraft-{project_name}"
+      env = lookup_environment(name)
+
+      if env and env.project_id == project_id:
+        return env
+
+      if not env or is_environment_reusable(env):
+        return create_fresh_environment(name)
+
+      name = f"snapcraft-{project_id}"
+      env = lookup_environment(name)
+      if is_environment_reusable(env) or ask_user_to_reuse(name):
+        return create_fresh_environment(name)
+
+      # User refused to re-use name, abort.
+      abort()
+
+    env = get_environment(project_path)
+#+END_EXAMPLE
+
+*** Removal
+    :PROPERTIES:
+    :CUSTOM_ID: removal
+    :END:
+
+Two options:
```

Comment:
Two options? Can you disambiguate this? I understand after reading everything, maybe this is two sections

Removal from withing project

Removal of 1...N projects